### DISCUSSION TITLE
CaseKeyPaths and the ForEachStore

### DISCUSSION THREAD

**Discussion by seanmrich** [pointfreeco/swift-composable-architecture#2584] [Q&A] [answered]
I created a toy project that manages color palettes using the new `Reducer` macro and `CaseKeyPath`s. I'm getting an error when scoping a `ForEachStore`. Here's the relevant reducers:
```swift
@Reducer
public struct PaletteEditor: Reducer {
  public struct State: Equatable {
    public var colors: IdentifiedArrayOf<PaletteColor.State>
  }
  
  public enum Action {
    case colors(IdentifiedActionOf<PaletteColor>)
  }
  
  public var body: some Reducer<State, Action> {
    Reduce { state, action in
      switch action {
      case .colors:
        return .none
      }
    }
    .forEach(\.colors, action: \.colors) {
      PaletteColor()
    }
  }
}

@Reducer
public struct PaletteColor {
  public struct State: Equatable, Identifiable {
    public var color: RGB
    public var id: ID
    
    public init(color: RGB, id: UUID? = nil) {
      self.color = color
      self.id = {
        if let id { return Tagged(id) }
        @Dependency(\.uuid) var uuid
        return Tagged(uuid())
      }()
    }
    
    public typealias ID = Tagged<Self, UUID>
  }
  
  public enum Action {}
  
  public var body: some Reducer<State, Action> {
    EmptyReducer()
  }
}
```

Then in the view for the palette, I get an error at the action keyPath: "Reference to member `colors` cannot be resolved without a contextual type"
```swift
public struct PaletteEditorView: View {
  let store: StoreOf<PaletteEditor>
  
  public var body: some View {
    ForEachStore(store.scope(state: \.colors, action: \.colors)) { colorStore in
    //                                                ^ Error 
      PaletteColorView(store: colorStore)
    }
  }
}
```

I tried fully qualifying the keyPath (perhaps wrongly) with `\PaletteEditor.Action.Cases.colors` but I just get "Failed to produce diagnostic" on the View. I referenced the 1.4 documentation on migrating to the new `CaseKeyPaths` as well as the `ForEachStore` but I'm not seeing where I went wrong. I'd sure appreciate any guidance you have.

**Comment by mbrandonw** [comment]
Hi @seanmrich, the tools you are wanting to use are not released in 1.4. They are still in beta, which you can find the details [here](https://github.com/pointfreeco/swift-composable-architecture/discussions/2565). If you want to try it out you will need to point your dependency on TCA to the branch `store-tree`. We will be releasing the tools in the next few days.

**Comment by seanmrich** [comment]
Well, that would explain it 🤦‍♂️.

Separately, if I turn on strict concurrency I get a warning about the action key path in `PaletteEditor.body`'s `forEach` method:
"Cannot form key path that captures non-sendable type 'KeyPath<PaletteEditor.Action.AllCasePaths, AnyCasePath<PaletteEditor.Action, IdentifiedAction<PaletteColor.State.ID, PaletteColor.Action>>>' (aka 'KeyPath<PaletteEditor.Action.AllCasePaths, AnyCasePath<PaletteEditor.Action, IdentifiedAction<Tagged<PaletteColor.State, UUID>, PaletteColor.Action>>>')"

I tried setting both `PaletteEditor.Action` and `PaletteColor.Action` to explicitly conform to `Sendable`, but it's the `KeyPath` itself that isn't `Sendable`. Looking at the code for the for each reducer, I don't see a `Sendable` requirement so I'm not sure what the warning is addressing.

**Reply by mbrandonw** [reply, marked as answer]
That is a known issue in the swift compiler and it should be fixed very soon. You will have to ignore that warning for now.

**Reply by seanmrich** [reply]
Excellent. Thanks for the guidance!
